edgeml_pytorch/graph/rnn.py

```diff
@@ -3,121 +3,198 @@
 import torch.nn as nn
+from torch.autograd import Function
 import numpy as np
 
-def gen_non_linearity(A, non_linearity):
+import edgeml_pytorch.utils as utils
+
+def onnx_exportable_rnn(input, fargs, cell, output):
+    class RNNSymbolic(Function):
+        @staticmethod
+        def symbolic(g, *fargs):
+            # NOTE: args/kwargs contain RNN parameters
+            return g.op(cell.name, *fargs, 
+                        outputs=1, hidden_size_i=cell.state_size,
+                        wRank_i=cell.wRank, uRank_i=cell.uRank,
+                        gate_nonlinearity_s=cell.gate_nonlinearity,
+                        update_nonlinearity_s=cell.update_nonlinearity)
+
+        @staticmethod
+        def forward(ctx, *fargs):
+            return output
+
+        @staticmethod
+        def backward(ctx, *gargs, **gkwargs):
+            raise RuntimeError("FIXME: Traced RNNs don't support backward")
+
+    return RNNSymbolic.apply(input, *fargs)
+
+def gen_nonlinearity(A, nonlinearity):
     '''
     Returns required activation for a tensor based on the inputs
 
-    non_linearity is either a callable or a value in
+    nonlinearity is either a callable or a value in
         ['tanh', 'sigmoid', 'relu', 'quantTanh', 'quantSigm', 'quantSigm4']
     '''
-    if non_linearity == "tanh":
+    if nonlinearity == "tanh":
         return torch.tanh(A)
-    elif non_linearity == "sigmoid":
+    elif nonlinearity == "sigmoid":
         return torch.sigmoid(A)
-    elif non_linearity == "relu":
+    elif nonlinearity == "relu":
         return torch.relu(A, 0.0)
-    elif non_linearity == "quantTanh":
+    elif nonlinearity == "quantTanh":
         return torch.max(torch.min(A, torch.ones_like(A)), -1.0 * torch.ones_like(A))
-    elif non_linearity == "quantSigm":
+    elif nonlinearity == "quantSigm":
         A = (A + 1.0) / 2.0
         return torch.max(torch.min(A, torch.ones_like(A)), torch.zeros_like(A))
-    elif non_linearity == "quantSigm4":
+    elif nonlinearity == "quantSigm4":
         A = (A + 2.0) / 4.0
         return torch.max(torch.min(A, torch.ones_like(A)), torch.zeros_like(A))
     else:
-        # non_linearity is a user specified function
-        if not callable(non_linearity):
-            raise ValueError("non_linearity is either a callable or a value " +
+        # nonlinearity is a user specified function
+        if not callable(nonlinearity):
+            raise ValueError("nonlinearity is either a callable or a value " +
                              + "['tanh', 'sigmoid', 'relu', 'quantTanh', " +
                              "'quantSigm'")
-        return non_linearity(A)
+        return nonlinearity(A)
 
 
-class BaseRNN(nn.Module):
-    '''
-    Generic equivalent of static_rnn in tf
-    Used to unroll all the cell written in this file
-    We assume data to be batch_first by default ie.,
-    [batchSize, timeSteps, inputDims] else
-    [timeSteps, batchSize, inputDims]
-    '''
+class RNNCell(nn.Module):
```

Comment:
I'm curious: do we have to do anything special to the existing cells, the ones already native to torch, to make it ONIX compatible ?

Reply:
I think they should be fine to export to ONNX without this ugly contraption.